text/3633-freeze-in-bounds.md

```diff
@@ -0,0 +1,239 @@
+
+`T: NoCell` is notably a requirement for static promotion (`const REF: &'a T;`) to be legal.
+
+Note that static promotion doesn't guarantee a single address: if `REF` is assigned to multiple variables,
+they may still refer to distinct addresses.
+
+Whether or not `T` implements `NoCell` may also affect whether `static STATIC: T` is placed
+in read-only static memory or writeable static memory, or the optimizations that may be performed
+in code that holds an immutable reference to `T`.
+
+# Semver hazard
+`NoCell` being an auto-trait that encodes a low level property of the types it is implemented for,
+you should avoid relying on external types maintaining that property, unless that
+contract is explicitly stated out-of-band (through documentation, for example).
+
+Conversely, authors that consider `NoCell` to be part of a type's contract should document this
+fact explicitly.
+
+## The ZST caveat
+While `UnsafeCell<T>` is currently `!NoCell` regardless of `T`, allowing `UnsafeCell<T>: NoCell` if `T` is
+a Zero-Sized-Type is currently under consideration.
+
+Therefore, the advised way to make your types `!NoCell` regardless of their actual contents is to add a
+[`PhantomCell`](core::marker::PhantomCell) field to it.
+
+[`PhantomData<T>`](core::marker::PhantomData) only implements `NoCell` if `T` does, making it a good way
+to conditionally remove a generic type's `NoCell` auto-impl.
+
+# Safety
+This trait is a core part of the language, it is just expressed as a trait in libcore for
+convenience. Do *not* implement it for other types.
+```
+
+Mention could be added to `UnsafeCell` and atomics that adding one to a previously `NoCell` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoke its implementation of `NoCell`.
+
+## Fixing `core::marker::PhantomData`'s `NoCell` impl
+
+At time of writing, `core::marker::PhantomData<T>` implements `NoCell` regardless of whether or not `T` does.
+
+This is now considered a bug, with the corrected behaviour being that `core::marker::PhantomData<T>` only implements `NoCell` if `T` does.
+
+While crates that would "observe" this change exist, the current consensus is that it would only break invalid usages of that invariable bound.
+Most crates that would observe this change could replace their usage of `core::marker::PhantomData<T>` by `core::marker::PhantomData<Ptr<T>>` where `Ptr<T>`
+is a pointer-type with the relevant caracteristics regarding lifetime, `Send` and `Sync`ness.
+
+The author doesn't have the necessary knowledge to implement this change, which should still be subject to a crater run to ensure no valid use-cases were missed.
+
+This behaviour change shall be introduced at the same time as the stabilization of `NoCell` in bounds.
+
+## `core::marker::PhantomCell`
+
+This ZST is proposed as a means for maintainers to reliably opt out of `NoCell` without constraining currently `!NoCell` ZSTs to remain so.
```

Comment:
Is there any situation where `PhantomCell` is a necessity, other than to prevent future API breakage beforehand? IIUC it's always required to use `UnsafeCell` to create internal mutability, thus it's still unsound to add a `_marker: PhantomCell` and expect to do anything creative with this struct (like mutating other fields behind `&`), yes?

If so, `NoCell`'s behavior sounds like a diverge from soundness guarantees and optimization invariants (`Freeze`). People who write `PhantomCell` would be only to opt-out their structs from const promotions, and they won't expect a surprising de-optimization of all other code using `&Struct`. Also in this way, user has no reason to override `Freeze` impl, making it unnecessary to be exposed to users at all.